storage/innobase/log/log0recv.cc

```diff
@@ -2463,6 +2519,121 @@ static byte *recv_parse_or_apply_log_rec_body(
+      size += block->total_size;
+      if (space.second->m_pages.size() >= 1) {
+        pages += space.second->m_pages.size();
+      }
+    }
+  }
+
+  return std::make_pair(size, pages);
+}
+
+static pxb_mem_block *pxb_add_new_block(pxb_space_page *space, ulint size) {
+  pxb_mem_block *block;
+  pxb_mem_block *new_block = new pxb_mem_block();
+  ulint new_size;
+  block = space->blocks.back();
+  new_size = 2 * block->len;
+  if (new_size > MEM_MAX_ALLOC_IN_BUF) {
+    new_size = MEM_MAX_ALLOC_IN_BUF;
+  }
+  if (new_size < size) {
+    new_size = size;
+  }
+  ulint len = MEM_BLOCK_HEADER_SIZE + MEM_SPACE_NEEDED(new_size);
+
+  if (len >= UNIV_PAGE_SIZE / 2) len = UNIV_PAGE_SIZE;
+
+  new_block->len = len;
+  new_block->total_size = block->total_size + len;
+  new_block->free = MEM_BLOCK_HEADER_SIZE;
+  space->blocks.push_back(new_block);
+
+  return new_block;
+}
+
+/** Get the page map for a tablespace. It will create one if one isn't found.
+@param[in]	space_id	Tablespace ID for which page map required.
+@return the space data  */
+static pxb_space_page *pxb_recv_get_page_map(space_id_t space_id) {
+  auto it = pxb_recv_sys->spaces->find(space_id);
+
+  if (it != pxb_recv_sys->spaces->end()) {
+    return (it->second);
+  }
+
+  ulint len = MEM_BLOCK_HEADER_SIZE + MEM_SPACE_NEEDED(256);
+  pxb_mem_block *block = new pxb_mem_block(len);
+  pxb_space_page *space_page = new pxb_space_page();
+  space_page->blocks.push_back(block);
+  using Space = pxb_recv_sys_t::pxb_space_page_t *;
+  using Value = pxb_recv_sys_t::Spaces::value_type;
+
+  auto where =
+      pxb_recv_sys->spaces->insert(it, Value({space_id, Space(space_page)}));
+
+  return (where->second);
+}
+
+static void recv_calculate_hash_heap(mlog_id_t type, space_id_t space_id,
+                                     page_no_t page_no, byte *body,
+                                     byte *rec_end, lsn_t start_lsn) {
+  if (!recv_recovery_on && xtrabackup_start_checkpoint > start_lsn) return;
+
+  ut_ad(type != MLOG_FILE_DELETE);
+  ut_ad(type != MLOG_FILE_CREATE);
+  ut_ad(type != MLOG_FILE_RENAME);
+  ut_ad(type != MLOG_FILE_EXTEND);
+  ut_ad(type != MLOG_DUMMY_RECORD);
+  ut_ad(type != MLOG_INDEX_LOAD);
+
+  /* check if we already have a Heap for this space id */
+  auto space = pxb_recv_get_page_map(space_id);
+
+  pxb_mem_block *last_block = space->blocks.back();
+  if (last_block->len < (last_block->free + MEM_SPACE_NEEDED(sizeof(recv_t)))) {
+    last_block = pxb_add_new_block(space, sizeof(recv_t));
+  }
+  last_block->free = last_block->free + MEM_SPACE_NEEDED(sizeof(recv_t));
+
+  if (std::find(space->m_pages.begin(), space->m_pages.end(), page_no) ==
+      space->m_pages.end()) {
+    if (last_block->len <
+        (last_block->free + MEM_SPACE_NEEDED(sizeof(recv_addr_t)))) {
+      last_block = pxb_add_new_block(space, sizeof(recv_addr_t));
+    }
+    last_block->free = last_block->free + MEM_SPACE_NEEDED(sizeof(recv_addr_t));
```

Comment:
I think they are independent from each other. Here we are doing what the function name says ( `calculate_hash_heap` 
) the core logic of the function is actually to calculate the amount of memory. The hard part to move it to add_new_block is that we need to keep traversing the log record (like on `while (rec_end > body)` ) . In the end it will result as add_new_block code been repeated every time we need to add a new block.